include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp

```diff
@@ -96,7 +96,7 @@ __device__ T __hipsycl_reduce_over_group(sub_group g, T x,
                                          BinaryOperation binary_op) {
   const size_t       lid        = g.get_local_linear_id();
   const size_t       lrange     = g.get_local_linear_range();
-  const unsigned int activemask = __activemask();
+  const unsigned int activemask = __ballot_sync(0xffffffff, 1);
```

Comment:
I'm not sure if
```
__ballot_sync(__activemask(), 1):
```
would be valid, they use something similar in their [blog](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/) at "Opportunistic Warp-level Programming".

If that isn't defined I personally am more comfortable with the original logic, maybe add a sync in front (which also doesn't guarantee the mask to be correct).
Just using `0xffffffff` for the active mask is in my opinion worse, since we would use the result of the shuffle which is also undefined, but the old version never failed the tests for me at least.

Since the spec expects us to also allow incomplete groups/sub_groups the only real fix (if __activemask() inside a *_sync function doesn't work) is to rewrite the functions to work without using the active mask. I'll check that option as well.

Reply:
> would be valid, they use something similar in their [blog](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/) at "Opportunistic Warp-level Programming".

I think they are using `__ballot_sync` **before** any divergence occurs, to get a mask of said divergence.

`__ballot_sync(mask, 1)` will either produce the mask itself (if the mask is correct) or, if the mask is too broad, will likely hang (although I think it's not officially defined). So, per my understanding `__ballot_sync(__activemask(), 1)` is the same as `__activemask()`.

> Since the spec expects us to also allow incomplete groups/sub_groups the only real fix (if __activemask() inside a *_sync function doesn't work) is to rewrite the functions to work without using the active mask. I'll check that option as well.

Perhaps [`coalesced_threads()`](https://developer.nvidia.com/blog/cooperative-groups/) can be the answer. But I have not read deeply about cooperative groups, so don't know if there are any pitfalls.

We can also check all the conditional happening before the current sub-group collective operation, and insert `__ballot_sync` before them to keep the mask updated. But that looks complicated and wasteful.

Long story short, having partial warps is very complicated. For starters, I suggest getting the full warps right.

>  Just using 0xffffffff for the active mask is in my opinion worse, since we would use the result of the shuffle which is also undefined, but the old version never failed the tests for me at least.

On CUDA, we're using the full mask for shuffle intrinsic:

https://github.com/illuhad/hipSYCL/blob/9e78270f5c5e85d3f030e565628edc33eee61f94/include/hipSYCL/sycl/libkernel/generic/hiplike/warp_shuffle.hpp#L90

https://github.com/illuhad/hipSYCL/blob/9e78270f5c5e85d3f030e565628edc33eee61f94/include/hipSYCL/sycl/libkernel/generic/hiplike/warp_shuffle.hpp#L146

- In the case of having only part of the warp, the old code will not work anyway (hanging at `__shfl_sync`). But it also potentially has a data race with a full warp.
- Using the full mask here will at least guarantee that the full warp case works fine, since we will be correctly applying the `binary_op`.
- Using `__ballot_sync` will also work, but it's more instructions with no benefit, as far as I can tell.

Reply:
I spend some more time reading up on how this could be solved.

I couldn't find any mention in the [programming guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-shuffle-functions) or the [PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-shfl-sync) that `0xFFFFFFFF` as mask leads to undefined behavior, as long as only data from active threads is used.

While __ballot_sync using `0xFFFFFFFF` as mask is undefined behavior if not all threads are active, `__match_any_sync` does not have this limitation (at least it isn't mentioned in the [programming guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-match-functions) or [PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-match-sync)).

So replacing `__ballot_sync` with `__match_any_sync` should solve our issue.

I need to test my code, if everything works I'll update the PR.

Reply:
> I couldn't find any mention in the [programming guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-shuffle-functions) or the [PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-shfl-sync) that 0xFFFFFFFF as mask leads to undefined behavior, as long as only data from active threads is used.

From https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/:

> Don’t just use FULL_MASK (i.e. 0xffffffff for 32 threads) as the mask value. If not all threads in the warp can reach the primitive according to the program logic, then using FULL_MASK may cause the program to hang.

Reply:
> Using FULL_MASK shouldn't be a problem for us, since the SYCL standard requires the group_functions to be run by all threads

Are we guaranteed to have a full warp? E.g., what if we have only 16 threads total? 

From the description of `sycl::sub_group::get_local_range()` it seems that such things are allowed:

> Return a range representing the size of the sub-group. This size may be less than the value returned by `get_max_local_range()`, depending on the position of the sub-group within its parent work-group and the manner in which sub-groups are constructed by the implemen­tation.

Reply:
We are not guaranteed to have a full warp, that's why we need to use the mask. According to the documentation shuffle and match instructions only synchronize threads non-exited threads.

> shfl.sync will cause executing thread to wait until all non-exited threads corresponding to membermask have executed shfl.sync with the same qualifiers and same membermask value before resuming execution.

The same is true for the match instruction. When looking at the PTX ISA even `__ballot_sync` might work

> vote.sync will cause executing thread to wait until all non-exited threads corresponding to membermask have executed vote.sync with the same qualifiers and same membermask value before resuming execution.

I interpreted the description in the programming guide differently before, but it only seems to be a problem if different masks or instructions are used.

Reply:
>  shfl.sync will cause executing thread to wait until all non-exited threads corresponding to membermask have executed shfl.sync with the same qualifiers and same membermask value before resuming execution.

Thanks, I missed the "non-exited" part. Then that should be fine; sorry for doubting.

> The same is true for the match instruction. When looking at the PTX ISA even __ballot_sync might work

Agree, `__ballot_sync` should be ok.

Reply:
It's so easy to miss or misread some condition for these low-level functions.

Happy you also had an eye on it.